src/framework/global/modularity/imoduleexport.h

```diff
@@ -24,28 +24,71 @@
 #define MU_MODULARITY_IMODULEEXPORT_H
 
 #include <memory>
-
-#define INTERFACE_ID(cls)               \
-public:                                 \
-    static const char* interfaceId() {  \
-        static const char* id = #cls;   \
-        return id;                      \
-    }                                   \
+#include "moduleinfo.h"
 
 namespace mu::modularity {
-class IModuleExportInterface
+struct InterfaceInfo {
+    std::string_view id;
+    std::string_view module;
+    bool internal = false;
+    constexpr InterfaceInfo(std::string_view i, std::string_view m, bool intr)
+        : id(i), module(m), internal(intr) {}
+};
+
+class IModuleInterface
+{
+public:
+    virtual ~IModuleInterface() = default;
+};
+
+class IModuleExportInterface : public IModuleInterface
+{
+public:
+    virtual ~IModuleExportInterface() = default;
+
+    static constexpr bool isInternalInterface() { return false; }
+};
+
+class IModuleInternalInterface : public IModuleInterface
+{
+public:
+    virtual ~IModuleInternalInterface() = default;
+
+    static constexpr bool isInternalInterface() { return true; }
+};
+
+class IModuleCreator
 {
 public:
-    virtual ~IModuleExportInterface() {}
+    virtual ~IModuleCreator() = default;
+    virtual std::shared_ptr<IModuleInterface> create() = 0;
 };
 
-struct IModuleExportCreator {
-    virtual ~IModuleExportCreator() {}
-    virtual std::shared_ptr<IModuleExportInterface> create() = 0;
+struct IModuleExportCreator : public IModuleCreator {
+    virtual ~IModuleExportCreator() = default;
+    static constexpr bool isInternalInterface() { return false; }
+};
+
+struct IModuleInternalCreator : public IModuleCreator {
+    virtual ~IModuleInternalCreator() = default;
+    static constexpr bool isInternalInterface() { return true; }
 };
 }
 
+#define INTERFACE_ID(id)                                                \
+public:                                                                 \
+    static constexpr mu::modularity::InterfaceInfo interfaceInfo() {    \
+        constexpr std::string_view sig(IOC_FUNC_SIG);                   \
+        constexpr mu::modularity::InterfaceInfo info(#id, mu::modularity::moduleNameBySig(sig), isInternalInterface());    \
```

Comment:
Wouldn't it be better to use `static constexpr` instead of just `constexpr`? I've heard that just `constexpr` gives no guarantee that it _is_ evaluated at compile time, only that it _can_ be. Using `static constexpr` gives a better guarantee.

Reply:
I get error  
```
/media/igor/G512/Dev/MuseScore/src/framework/global/modularity/imoduleexport.h:81: error: ‘sig’ declared ‘static’ in ‘constexpr’ function
In file included from /media/igor/G512/Dev/MuseScore/src/framework/global/io/internal/../ifilesystem.h:25,
                 from /media/igor/G512/Dev/MuseScore/src/framework/global/io/internal/filesystem.h:25,
                 from /media/igor/G512/Dev/MuseScore/src/framework/global/io/internal/filesystem.cpp:22,
                 from /media/igor/G512/Dev/MuseScore/build.debug/src/framework/global/CMakeFiles/global.dir/Unity/unity_4_cxx.cxx:3:
/media/igor/G512/Dev/MuseScore/src/framework/global/io/internal/../ifilesystem.h: In static member function ‘static constexpr mu::modularity::InterfaceInfo mu::io::IFileSystem::interfaceInfo()’:
/media/igor/G512/Dev/MuseScore/src/framework/global/modularity/imoduleexport.h:81:43: error: ‘sig’ declared ‘static’ in ‘constexpr’ function
   81 |         static constexpr std::string_view sig(IOC_FUNC_SIG);                   \
      |                                           ^~~
/media/igor/G512/Dev/MuseScore/src/framework/global/io/internal/../ifilesystem.h:35:5: note: in expansion of macro ‘INTERFACE_ID’
   35 |     INTERFACE_ID(IFileSystem)
      |     ^~~~~~~~~~~~
```

Reply:
Oh. Then never mind :)